### PR TITLE
Add Message-ID Tracking on receipts (WIP ish?)

### PR DESCRIPTION
**Pull request by rposborne**
Refactor MailDispatcher to consume receipts
Extract RecipentFilter into own class.

**Warning Side effects** 
Reciepts are not made if a user opts out. 
Fully drops support for passing an array to the mailer.

### CODE REVIEW

**Comment by rposborne** [comment]
So this applies to #375 Had to go on top of the shoulda matchers to get tests working. 

The only real debate is my laziness around trying to figure out the right behavior for send array to mailer.  I am not sure the number of people using this behavior?  But wanted to put this out here so maybe we can refactor.

**Comment by rposborne** [comment]
@jcoyne ping here too :)  This one isn't cut and dry.

**Comment by jcoyne** [comment]
To start with, this needs to be rebased against master.

File: lib/mailboxer/mail_dispatcher.rb [hunk 1]
```diff
@@ -1,21 +1,18 @@
...
-    attr_reader :mailable, :recipients
-
-    def initialize(mailable, recipients)
-      @mailable, @recipients = mailable, recipients
+    def initialize(mailable, receipts)
+      @mailable, @receipts = mailable, receipts
     end
 
     def call
+      # Why is this here, the dispatch should dispatch email not decide to dispatch
```
Comment (jcoyne) [inline lib/mailboxer/mail_dispatcher.rb]: Please make an issue or a PR for this.  This sort of comment gets confusing in the code.

**Comment by jcoyne** [comment]
Why are you removing `mailer_wants_array` (I don't know what that is for myself)?  It seems like that may make this PR break backwards compatibility.

**Comment by rposborne** [comment]
@jcoyne yes it will break something.   The point is to be able to have a record in the DB that we can update when we "verify" that an email has been read (i.e. postmark web hook).  To be able to correlate this information we need a message-id from the action mailer to be persisted somewhere.   

The array issue is that we only get one message-id.   It's not insurmountable but I wanted to have a discussion about the feature set before pursuing it.   

I assume for the behavior in the array mailer we just need to be clear that you loose resolution of individuals reading the message.

**Comment by rposborne** [comment]
Just a heads up we are using this in production and things are running quite nicely.

**Comment by rposborne** [comment]
@jcoyne do you have any thoughts on this.   We are tossing up a pivot to ahoy_email in favor of broad ability to track email (including mailboxer)

**Comment by jcoyne** [comment]
@rposborne Sorry, i'd been overlooking this because the title says "WIP". Is it now ready?

**Comment by rposborne** [comment]
@jcoyne The only thing that would be WIP, is if we don't want to break compatibility with the passing an array to mailers.  I would be for leaving support for that out has it has some weird read state behavior.

**Pull request merged**